File: Find Kth Bit in Nth Binary String.py
class Solution:
    def findKthBit(self, n: int, k: int) -> str:
        def recur(m, j):
            # print(m, j)
            if j == 1:
                return 0

            x = 2 ** (m - 1)
            if j == x:
                return 1
            if j > x:
                return not recur(m - 1, 2 * x - j)

            y = j
            for i in range(1, m):
                minus = 2 ** (i - 1)
                if y - minus <= 0:
                    return recur(i, j)
                y -= minus
            return 0

        return str(int(recur(n, k)))

File: Find Kth Bit in Nth Binary String.cpp
class Solution {
public:
    bool recur(int m, int j) {
        if (j == 1) return false;

        int x = 1 << (m-1);
        if (j == x) return true;
        if (j > x) return !recur(m-1, 2*x-j);

        int y = j;
        for (int i = 1; i < m; i++) {
            int minus = 1 << (i-1);
            if (y - minus <= 0) return recur(i, j);
            y -= minus;
        }
        return 0;
    }

    char findKthBit(int n, int k) {
        return char((int) recur(n, k) + 48);
    }
};
